### PR TITLE
color: add hex2color, color2hex, rgb2str, str2rgb

### PR DESCRIPTION
**Pull request by cgundogan**
I needed some conversions between colors given in hex code (string) and the `rgb_color_t` or `uint32_t` forms.
I thought sharing this might be a good idea.

### CODE REVIEW

File: sys/color/color.c [hunk 1]
```diff
@@ -127,3 +127,15 @@ void color_hsv2rgb(color_hsv_t *hsv, color_rgb_t *rgb)
             break;
     }
 }
+
+void color_hex2rgb(uint32_t hex, color_rgb_t *rgb)
+{
+    rgb->r = ((hex >> 16) & 0xFF);
+    rgb->g = ((hex >> 8) & 0xFF);
+    rgb->b = (hex & 0xFF);
```
Comment (authmillenon) [inline sys/color/color.c]: Hacky, but more afficienent way could be

```C
uint32_t *ptr = (uint32_t *)rgb;
*ptr = hex << 8;
```

(please test this if you decide to use this. Might be, that I'm wrong about the shift.

File: sys/color/color.c [hunk 2]
```diff
@@ -127,3 +127,15 @@ void color_hsv2rgb(color_hsv_t *hsv, color_rgb_t *rgb)
...
+}
+
+void color_rgb2hex(color_rgb_t *rgb, uint32_t *hex)
+{
+    *hex = (rgb->r << 16) | (rgb->g << 8) | (rgb->b);
```
Comment (authmillenon) [inline sys/color/color.c]: The reverse of above could be applied here

File: sys/color/color.c (hunk 1 above)
Comment (cgundogan) [inline sys/color/color.c]: what if the struct has padding?

**Comment by authmillenon** [comment]
Can you add a description about the intended use? It's not very clear to me.

**Comment by cgundogan** [comment]
@authmillenon I added a description. I also amended `str2rgb` and `rgb2str` conversions, because I needed them for a demo. Furthermore, I introduced a very small unittest for checking the functionality of `rgb2str` and `str2rgb`

**Comment by kaspar030** [comment]
Now that you're touching this, why don't we save a colour as 32bit integer including alpha?

**Comment by cgundogan** [comment]
@kaspar030 can you have a look at the last commit? Is that what you had in mind?

File: sys/include/color.h [hunk 3]
```diff
@@ -28,13 +29,17 @@ extern "C" {
...
+ * @brief Data-structure describing a ARGB color
  */
-typedef struct {
-    uint8_t r;          /**< red value      [0 - 255] */
-    uint8_t g;          /**< green value    [0 - 255] */
-    uint8_t b;          /**< blue value     [0 - 255] */
-} color_rgb_t;
+typedef union {
+    uint32_t raw;
+    struct __attribute__((packed)) {
```
Comment (cgundogan) [inline sys/include/color.h]: @kaspar030 do you think this attribute is superfluous here? I was not sure whether a 64bit platform e.g. might put padding into the argb struct (non-commit comment version)

File: tests/unittests/tests-color/tests-color.c [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    TEST_ASSERT_EQUAL_INT(0x9A, color.argb.g);
+    TEST_ASSERT_EQUAL_INT(0x1D, color.argb.b);
+}
+
+static void test_str2rgb__fail(void)
+{
+    const char *color_str = "F09A1D";
+    color_argb_t color;
+
+    TEST_ASSERT_EQUAL_INT(-1, color_str2rgb(color_str, &color));
```
Comment (authmillenon) [inline tests/unittests/tests-color/tests-color.c]: What does this fail?

Comment (cgundogan) [inline tests/unittests/tests-color/tests-color.c]: it fails because the string does not start with an `#` sign

**Comment by cgundogan** [comment]
@kaspar030 removed the packed attribute from the `argb_t` struct

File: tests/unittests/tests-color/tests-color.c (hunk 4 above)
Comment (authmillenon) [inline tests/unittests/tests-color/tests-color.c]: Why is it required? Can't you save a few byte of ROM by just ignoring every non hex-char?

Comment (cgundogan) [inline tests/unittests/tests-color/tests-color.c]: yeah, you are right.

**Comment by cgundogan** [comment]
addressed @authmillenon's comments and squashed immediately.

**Comment by cgundogan** [comment]
@authmillenon are you OK with the current state?

**Comment by cgundogan** [comment]
@haukepetersen mentioned some concerns regarding the alpha value that this PR introduces. Could mention them here again?

**Comment by haukepetersen** [comment]
@kaspar030: why should we want an alpha channel here? For simple stuff like LED stripes this is not needed - and so why don't we keep it as simple as possible?

**Comment by cgundogan** [comment]
I removed the last commit from this PR, which introduced the alpha value.
The commit can be found in [1], in case the discussion keeps rolling.

@haukepetersen any objections to merge once travis is green?

[1] https://github.com/cgundogan/RIOT/tree/color_backup

**Comment by haukepetersen** [comment]
One thing that might prove useful is to allow also lower-case letters when parsing strings (e.g. '#caffee'), but this we can add later, also.

ACK once you made Travis happy.

**Comment by cgundogan** [comment]
thats possible currently, because the characters get transformed to upper case with `(str[0] &~ 0x20)`. Maybe I should add a testcase for that later

**Comment by haukepetersen** [comment]
nice, overlooked that one... Merge at will once Travis agrees.

**Comment by kaspar030** [comment]
> @kaspar030: why should we want an alpha channel here? For simple stuff like LED stripes this is not needed - and so why don't we keep it as simple as possible?

Unless using multiple values or packed structs, it doesn't make a difference.

**Comment by cgundogan** [comment]
added a unittest for the lower case str2rgb and fixed some travis complains

**Comment by cgundogan** [comment]
travis is happy. @kaspar030 @haukepetersen I will merge this for now as this will be super handy for the demo. I will open a follow-up PR with the addition of the alpha value later so that the discussion can be continued.

**Pull request merged**